crates/agent/src/api/authorize.rs

```diff
@@ -0,0 +1,190 @@
+
+        if jsonwebtoken::decode::<proto_gazette::Claims>(token, &key, &validation).is_ok() {
+            verified = true;
+            break;
+        }
+    }
+    if !verified {
+        anyhow::bail!("no data-plane keys validated against the token signature");
+    }
+
+    // Query for a task => collection pair and their RBAC authorization.
+    let (task_name, collection_name, collection_data_plane_id, mut authorized) =
+        agent_sql::data_plane::verify_task_authorization(
+            &app.pg_pool,
+            task_shard,
+            journal_name_or_prefix,
+            required_role,
+        )
+        .await?
+        .unwrap_or((
+            String::new(),
+            models::Collection::default(),
+            models::Id::zero(),
+            false,
+        ));
+
+    // As a special case outside of the RBAC system, allow a task to write
+    // to its designated partition within its ops collections.
+    if !authorized
+        && required_role == "write"
+        && (collection_name == task_data_plane.ops_logs_name
+            || collection_name == task_data_plane.ops_stats_name)
+        && journal_name_or_prefix.ends_with(&format!(
+            "/kind={task_type}/name={}/pivot=00",
+            labels::percent_encoding(&task_name).to_string(),
+        ))
+    {
+        authorized = true;
+    }
+
+    if !authorized {
+        let ops_suffix = format!(
+            "/kind={task_type}/name={}/pivot=00",
+            labels::percent_encoding(&task_name).to_string(),
+        );
+        tracing::warn!(
+            %task_type,
+            %task_shard,
+            %journal_name_or_prefix,
+            required_role,
+            ops_logs=%task_data_plane.ops_logs_name,
+            ops_stats=%task_data_plane.ops_stats_name,
+            %ops_suffix,
+            "task authorization rejection context"
+        );
+        anyhow::bail!("task shard {task_shard} is not authorized to {journal_name_or_prefix} for {required_role}");
+    }
+    // We've now completed AuthN and AuthZ checks and can proceed.
+
+    // TODO(johnny): We can avoid a DB query in the common case,
+    // where a task and collection data-plane are the same.
+    // I'm not doing this yet to keep the code path simpler while we're testing.
+
+    let collection_data_plane = agent_sql::data_plane::fetch_data_planes(
+        &app.pg_pool,
+        vec![collection_data_plane_id],
+        "", // No default name to retrieve.
+        uuid::Uuid::nil(),
+    )
+    .await?
+    .pop()
+    .context("collection data-plane does not exist")?;
+
+    claims.iss = collection_data_plane.data_plane_fqdn;
+    claims.iat = jsonwebtoken::get_current_timestamp();
+    claims.exp = claims.iat + 3_600; // One hour.
+
+    let key = jsonwebtoken::EncodingKey::from_base64_secret(
+        collection_data_plane
+            .hmac_keys
+            .first()
+            .map(String::as_str)
+            .unwrap_or_default(),
```

Comment:
nit: seems like an empty string here might be bad somehow. Should it return an error?